README.rst
yet-another-django-profiler README
==================================

.. image:: https://travis-ci.org/safarijv/yet-another-django-profiler.svg?branch=master
    :target: https://travis-ci.org/safarijv/yet-another-django-profiler

.. image:: https://coveralls.io/repos/safarijv/yet-another-django-profiler/badge.svg
    :target: https://coveralls.io/r/safarijv/yet-another-django-profiler

.. image:: https://pypip.in/py_versions/yet-another-django-profiler/badge.svg
    :target: https://pypi.python.org/pypi/yet-another-django-profiler/
    :alt: Supported Python versions

.. image:: https://pypip.in/implementation/yet-another-django-profiler/badge.svg
    :target: https://pypi.python.org/pypi/yet-another-django-profiler/
    :alt: Supported Python implementations

.. image:: https://pypip.in/license/yet-another-django-profiler/badge.svg
    :target: https://pypi.python.org/pypi/yet-another-django-profiler/
    :alt: License

.. image:: https://pypip.in/download/yet-another-django-profiler/badge.svg
    :target: https://pypi.python.org/pypi//yet-another-django-profiler/
    :alt: Downloads

.. image:: https://pypip.in/status/yet-another-django-profiler/badge.svg
    :target: https://pypi.python.org/pypi/yet-another-django-profiler/
    :alt: Development Status

.. image:: https://pypip.in/wheel/yet-another-django-profiler/badge.svg
    :target: https://pypi.python.org/pypi/yet-another-django-profiler/
    :alt: Wheel Status

.. image:: https://pypip.in/egg/yet-another-django-profiler/badge.svg
    :target: https://pypi.python.org/pypi/yet-another-django-profiler/
    :alt: Egg Status

Yet Another Django Profiler attempts to combine the best features of assorted
other Django profiling utilities that have been created over the years.
(For more background information, see my
`blog post <http://blog.safariflow.com/2013/11/21/profiling-django-via-middleware/>`_
on the topic.)

Installation
------------
First, get the code via pip install::

    pip install yet-another-django-profiler

Then add ``yet_another_django_profiler.middleware.ProfilerMiddleware`` to your
``MIDDLEWARE_CLASSES`` Django setting (typically at the end of the list, if
you want to include profiling data on the other middleware that's in use).
If you want to generate call graphs with the middleware, you also need to
install `Graphviz <http://www.graphviz.org/Download.php>`_.  If you want to
use the "profile" management command, you'll also need to add
``yet_another_django_profiler`` to the ``INSTALLED_APPS`` setting.

Middleware Usage
----------------
The simplest usage is to just add a ``profile`` parameter to the URL of a
Django view.  This uses Graphviz to generate a PDF representation of the call
graph for the code executed to perform the view, and returns that as the
response to the request instead of the rendered view itself.  So calling a
URL like ``http://localhost:8000/admin/?profile`` shows a PDF like
`this <https://github.com/safarijv/yet-another-django-profiler/blob/master/docs/admin_call_graph.pdf?raw=true>`_
in the browser.

Alternatively, you can display a table of called functions ordered by the
desired statistic by using a URL such as ``http://localhost:8000/?profile=time``.
The available sorting options are:

* ``calls`` (call count)

* ``cumulative`` (cumulative time)

* ``file`` (file name, same as ``module``)

* ``module`` (file name, same as ``file``)

* ``pcalls`` (primitive call count)

* ``line`` (line number)

* ``name`` (function name)

* ``nfl`` ( function name/file/line)

* ``stdname`` (standard name)

* ``time`` (internal time)

By default, only the top 20% of function calls are included in the table.  To
change that, add a ``fraction`` parameter with the desired display ratio
(hence the default value is ``fraction=.2``).  Alternatively, you can
instead specify a maximum number of function calls to display using the
``max_calls`` parameter.  And if you specify a regular expression with the
``pattern`` parameter, only calls of functions whose names match the
specified pattern will be displayed.  (I'd recommend sticking to basic
sub-strings unless you really enjoy figuring out how to URL-escape special
characters.)

If you forget the available sorting options and such, you can use
``profile=help`` as a request parameter to display the usage instructions in
the browser.

Management Command Usage
------------------------
yet-another-django-profiler includes a ``profile`` management command which can
be used to profile other Django management commands::

    Usage: django-admin.py profile [options] other_command <argument argument ...>

    Profile another Django management command

    Options:
      -v VERBOSITY, --verbosity=VERBOSITY
                            Verbosity level; 0=minimal output, 1=normal output,
                            2=verbose output, 3=very verbose output
      --settings=SETTINGS   The Python path to a settings module, e.g.
                            "myproject.settings.main". If this isn't provided, the
                            DJANGO_SETTINGS_MODULE environment variable will be
                            used.
      --pythonpath=PYTHONPATH
                            A directory to add to the Python path, e.g.
                            "/home/djangoprojects/myproject".
      --traceback           Raise on exception
      -o PATH, --output=PATH
                            Path to a file in which to store the profiling output
                            (required if generating a call graph PDF, other
                            results are output to the console by default)
      -s SORT, --sort=SORT  Statistic by which to sort the profiling data (default
                            is to generate a call graph PDF instead)
      -f FRACTION, --fraction=FRACTION
                            The fraction of total function calls to display (the
                            default of .2 is omitted if max-calls or pattern are
                            specified)
      -m MAX_CALLS, --max-calls=MAX_CALLS
                            The maximum number of function calls to display
      -p PATTERN, --pattern=PATTERN
                            Regular expression filter for function display names
      -b BACKEND, --backend=BACKEND
                            Profiler backend to use (cProfile or yappi)
      --version             show program's version number and exit
      -h, --help            show this help message and exit

Sample usage:

* ``django-admin.py profile -s time test --failfast my_app/my_module.py:TestClass.test_function``
* ``django-admin.py profile -o ~/Downloads/call_graph.pdf collectstatic``

Settings
--------
The middleware is designed to be available whenever the ``DEBUG`` setting is
True, and removes itself from the middleware chain otherwise (so it can safely
be left in the dependencies for production deployments without performance or
security problems).  If for some reason you want to change this behavior, you
can set the ``YADP_ENABLED`` boolean setting directly to determine whether the
middleware is active or not.

If you have pages where the default profiling parameter names conflict with
existing parameters in the application, you can choose different ones via the
following settings:

* ``YADP_PROFILE_PARAMETER`` (default is "profile")

* ``YADP_FRACTION_PARAMETER`` (default is "fraction")

* ``YADP_MAX_CALLS_PARAMETER`` (default is "max_calls")

* ``YADP_PATTERN_PARAMETER`` (default is "pattern")

You can use Yappi (`Yet Another Python Profiler <https://code.google.com/p/yappi/>`_)
as a profiler backend instead of cProfile. To do that just specify
``YADP_PROFILER_BACKEND = 'yappi'`` in the settings.  Note that Yappi does not
currently work on PyPy or CPython 3.2.

An effort is made to convert the absolute Python file paths provided by the
profiler to full-qualified module names (which are typically shorter and
easier to understand at a glance).  The default rules should work in most cases
but can be customized via the following settings:

* ``YADP_MODULE_PARENT_DIR_PATTERNS`` is a list of regular expression patterns.
  Everything in a module path up to and including a match of one of these
  patterns is removed from statistic tables and call graphs.  The default list
  is ``[r'\.egg[/\\]', r'site-packages[/\\]', r'python\d+\.\d+[/\\]']``.  The
  absolute path of the current working directory is also pruned.

* If the previous setting doesn't allow sufficient customization for your
  needs, the ``YADP_PATH_TO_MODULE_FUNCTION`` setting can be used to completely
  replace the function used for this task.  It should be the fully qualified
  name of your custom function, which takes an absolute file path as input and
  returns what you want to appear in the profiling output to represent that
  path.

In order to get simple and meaningful profiling data, a
`few other changes <https://github.com/safarijv/yet-another-django-profiler/blob/master/docs/settings.rst>`_
to your settings may be in order.

Running Tests
-------------
To run tests in all currently supported combinations of Python and Django, run
``tox``.  If you're running tox from a Python 2 environment, you can instead
run ``detox`` to execute all the test environments in parallel.  See the
`tox documentation <https://tox.readthedocs.org/en/latest/>`_ for instructions
on running a single test case or environment.

Internationalization
--------------------
Translations of text that can appear in the profiling results pages are managed
on `Transifex <https://www.transifex.com/projects/p/yet-another-django-profiler/>`_.
Feel free to request to be added as translator for a not-yet-supported language.
Django recommends not translating management command text for
`assorted technical reasons <https://docs.djangoproject.com/en/1.8/howto/custom-management-commands/#management-commands-and-locales>`_,
so those phrases currently aren't included.

For development tasks involving the translations (uploading message changes to
Transifex or fetching the latest translations from it), use
`transifex-client <http://docs.transifex.com/guides/client>`_.  By default, pip
installs a rather old stable version so you may want to specify a newer one::

    pip install transifex-client==0.11b3

When running the makemessages or compilemessages management commands, do so
from the ``yet_another_django_profiler`` directory.

License
-------
Due to gprof2dot being licensed under the LGPL v3, that's the license that
applies to this package as a whole.  However, the rest of the source files are
individually licensed under a more permissive 3-clause BSD license (so it is
possible to assemble a BSD-licensed package that omits only the call graph
generation feature).


docs/CHANGELOG.rst
yet-another-django-profiler Changelog
=====================================

1.1.1 (2017-08-31)
------------------
* Updating dependencies
* Added support for Django 1.11

1.1.0 (2017-01-10)
------------------
* Added support for Django 1.10 (including the new ``MIDDLEWARE`` setting)
* Dropped support for Django 1.5 and 1.6
* Fixed profiling of management commands with arguments in Django 1.8+
* Redirect stderr of the profiled command to be the same as the ``profile``
  command's stderr

1.0.3 (2016-02-08)
------------------
* Fixed profiling of Django admin pages (by removing profiling parameters
  before executing the view)
* Convert the PYTHONPATH to absolute paths for the subprocess used to run
  gprof2dot when generating call graphs (fixes problems loading setting files,
  etc.)
* Correctly profile views which return a StreamingHttpResponse (up through
  creation of the response, not including iteration over the content)

1.0.2 (2015-12-08)
------------------
* Support for Django 1.9

1.0.1 (2015-08-15)
------------------
* Add URL and command line parameter for yappi clock type

1.0.0 (2015-07-17)
------------------
* Support for Django 1.8
* Support for Python 3 and PyPy
* Option to use Yappi instead of cProfile as the profiler
* Internationalization (English and Japanese for now, other languages welcome)

0.3.0 (2015-02-23)
------------------
* Added "profile" management command for profiling other management commands
* Attempt to resolve absolute paths of Python modules to fully-qualified module
  names for more readable results, and added 2 new settings to customize this
* Upgraded gprof2dot.py (still bundled despite a recent version now being
  available on PyPI because it is slightly modified to use the new module name
  derivation code)
* Work around bug in Python >= 2.7.4 that doesn't recognize "file" as a valid
  sorting option (even though every other initial substring of "filename" is)

0.2.0 (2015-02-12)
------------------
Don't require particular Django versions (so pip won't try to auto-adjust it).

0.1.0 (2014-02-27)
------------------
Initial release


test_urls.py
# encoding: utf-8
# Created by Jeremy Bowman on Fri Feb 21 14:28:40 EST 2014
# Copyright (c) 2014, 2015 Safari Books Online. All rights reserved.
#
# This software may be modified and distributed under the terms
# of the 3-clause BSD license.  See the LICENSE file for details.
"""
URL configuration for Yet Another Django Profiler tests.
"""

import django
from django.conf.urls import include, url
from django.contrib import admin
from django.http import HttpResponse, StreamingHttpResponse
from django.views.generic.base import View


class TestView(View):

    def get(self, request, *args, **kwargs):
        return HttpResponse('This is only a test.')


class StreamingView(View):

    def get(self, request, *args, **kwargs):
        return StreamingHttpResponse(('Line {}\n'.format(i + 1) for i in range(100)))


if django.VERSION[0] == 1 and django.VERSION[1] < 7:
    admin.autodiscover()

urlpatterns = [
    url(r'^test/$', TestView.as_view(), name='test'),
    url(r'^test/streaming/$', StreamingView.as_view(), name='test_streaming'),
    url(r'^admin/', include(admin.site.urls)),
]


test_settings.py
# encoding: utf-8
# Created by Jeremy Bowman on Fri Feb 21 12:31:31 EST 2014
# Copyright (c) 2014, 2015 Safari Books Online. All rights reserved.
#
# This software may be modified and distributed under the terms
# of the 3-clause BSD license.  See the LICENSE file for details.
"""
Django settings for Yet Another Django Profiler tests.

For more information on this file, see
https://docs.djangoproject.com/en/1.6/topics/settings/

For the full list of settings and their values, see
https://docs.djangoproject.com/en/1.6/ref/settings/
"""

import os
from uuid import uuid4

import django

BASE_DIR = os.path.dirname(os.path.dirname(__file__))

SECRET_KEY = str(uuid4())
DEBUG = True

# Application definition

INSTALLED_APPS = (
    'django.contrib.contenttypes',
    'django.contrib.auth',
    'django.contrib.admin',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    'yet_another_django_profiler',
)

if django.VERSION[0] == 1 and django.VERSION[1] < 10:
    MIDDLEWARE_CLASSES = (
        'django.contrib.sessions.middleware.SessionMiddleware',
        'django.middleware.common.CommonMiddleware',
        'django.middleware.csrf.CsrfViewMiddleware',
        'django.contrib.auth.middleware.AuthenticationMiddleware',
        'django.contrib.messages.middleware.MessageMiddleware',
        'django.middleware.clickjacking.XFrameOptionsMiddleware',
        'yet_another_django_profiler.middleware.ProfilerMiddleware',
    )
else:
    MIDDLEWARE = (
        'django.contrib.sessions.middleware.SessionMiddleware',
        'django.middleware.common.CommonMiddleware',
        'django.middleware.csrf.CsrfViewMiddleware',
        'django.contrib.auth.middleware.AuthenticationMiddleware',
        'django.contrib.messages.middleware.MessageMiddleware',
        'django.middleware.clickjacking.XFrameOptionsMiddleware',
        'yet_another_django_profiler.middleware.ProfilerMiddleware',
    )

ROOT_URLCONF = 'test_urls'

if django.VERSION[0] == 1 and django.VERSION[1] < 8:
    TEMPLATE_DEBUG = False

    TEMPLATE_CONTEXT_PROCESSORS = (
        'django.contrib.auth.context_processors.auth',
        'django.core.context_processors.i18n',
        'django.contrib.messages.context_processors.messages',
    )

TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'APP_DIRS': True,
        'OPTIONS': {
            'context_processors': [
                'django.contrib.auth.context_processors.auth',
                'django.contrib.messages.context_processors.messages',
            ],
            'debug': False,
        }
    },
]

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',
        'NAME': os.path.join(BASE_DIR, 'test_db.sqlite3'),
    }
}

TIME_ZONE = 'UTC'
USE_L10N = True
USE_TZ = True
STATIC_URL = '/static/'


yet_another_django_profiler/tests/test_parameters.py
# encoding: utf-8
# Created by Jeremy Bowman on Fri Feb 21 17:28:37 EST 2014
# Copyright (c) 2014, 2015 Safari Books Online. All rights reserved.
#
# This software may be modified and distributed under the terms
# of the 3-clause BSD license.  See the LICENSE file for details.
"""
Yet Another Django Profiler request parameters tests
"""

from __future__ import unicode_literals

import platform
import sys

from django.core.urlresolvers import reverse
from django.test import TestCase
from django.test.utils import override_settings
from django.utils.text import force_text

import pytest

HELP_EXCERPT = 'profiling middleware'


class ParameterCases(object):
    """Parameter tests to be run for each profiler backend"""

    def test_call_graph(self):
        """Using "profile" without a parameter should yield a PDF call graph"""
        response = self._get_test_page('profile')
        assert response['Content-Type'] == 'application/pdf'

    def test_calls_by_count(self):
        """Using profile=calls should show a table of function calls sorted by call count"""
        response = self._get_test_page('profile=calls')
        self.assertContains(response, 'Ordered by: call count')

    def test_calls_by_cumulative(self):
        """Using profile=cumulative should show a table of function calls sorted by cumulative time"""
        response = self._get_test_page('profile=cumulative')
        self.assertContains(response, 'Ordered by: cumulative time')

    def test_calls_by_file_name(self):
        """Using profile=file should show a table of function calls sorted by file name"""
        response = self._get_test_page('profile=file')
        self.assertContains(response, 'Ordered by: file name')

    def test_calls_by_function_name(self):
        """Using profile=name should show a table of function calls sorted by function name"""
        response = self._get_test_page('profile=name')
        self.assertContains(response, 'Ordered by: function name')

    def test_calls_by_function_name_file_and_line(self):
        """Using profile=nfl should show a table of function calls sorted by function name, file, and line"""
        response = self._get_test_page('profile=nfl')
        self.assertContains(response, 'Ordered by: name/file/line')

    def test_calls_by_line_number(self):
        """Using profile=line should show a table of function calls sorted by line_number"""
        response = self._get_test_page('profile=line')
        self.assertContains(response, 'Ordered by: line number')

    def test_calls_by_module(self):
        """Using profile=module should show a table of function calls sorted by file name"""
        response = self._get_test_page('profile=module')
        self.assertContains(response, 'Ordered by: file name')

    def test_calls_by_primitive_call_count(self):
        """Using profile=pcalls should show a table of function calls sorted by primitive call count"""
        response = self._get_test_page('profile=pcalls')
        self.assertRegexpMatches(force_text(response.content, 'utf-8'), r'Ordered by: (primitive )?call count')

    def test_calls_by_stdname(self):
        """Using profile=stdname should show a table of function calls sorted by standard name"""
        response = self._get_test_page('profile=stdname')
        self.assertContains(response, 'Ordered by: standard name')

    def test_calls_by_time(self):
        """Using profile=time should show a table of function calls sorted by internal time"""
        response = self._get_test_page('profile=time')
        self.assertContains(response, 'Ordered by: internal time')

    def test_calls_by_clock(self):
        """Using profile=time should show a table of function calls sorted by internal time"""
        response = self._get_test_page('profile=time&clock=wall')
        self.assertContains(response, 'Ordered by: internal time')

    def test_help(self):
        """Using profile=help should yield usage instructions"""
        response = self._get_test_page('profile=help')
        self.assertContains(response, HELP_EXCERPT)

    def test_default_fraction(self):
        """By default, the fraction of displayed function calls should be 0.2"""
        response = self._get_test_page('profile=time')
        self.assertContains(response, 'due to restriction <0.2>')

    def test_custom_fraction(self):
        """It should be possible to specify the fraction of displayed function calls"""
        response = self._get_test_page('profile=time&fraction=0.3')
        self.assertContains(response, 'due to restriction <0.3>')

    def test_max_calls(self):
        """It should be possible to specify the maximum number of displayed function calls"""
        response = self._get_test_page('profile=time&max_calls=5')
        self.assertContains(response, 'to 5 due to restriction <5>')

    def test_pattern(self):
        """It should be possible to specify a regular expression filter pattern"""
        response = self._get_test_page('profile=time&pattern=test')
        self.assertRegexpMatches(force_text(response.content, 'utf-8'), r"due to restriction <u?'test'>")

    def test_streaming_call_graph(self):
        """It should be possible to get a PDF call graph for a view which returns a StreamingHttpResponse"""
        response = self._get_streaming_page('profile')
        assert response['Content-Type'] == 'application/pdf'

    def test_streaming_text(self):
        """It should be possible to get text profiling data for a view which returns a StreamingHttpResponse"""
        response = self._get_streaming_page('profile=time')
        self.assertContains(response, 'Ordered by: internal time')

    def _get_streaming_page(self, params=''):
        url = reverse('test_streaming')
        if params:
            url += '?' + params
        return self.client.get(url)

    def _get_test_page(self, params=''):
        url = reverse('test')
        if params:
            url += '?' + params
        return self.client.get(url)


@override_settings(YADP_ENABLED=True)
class CProfileTest(TestCase, ParameterCases):
    """Profiling parameter tests using cProfile"""

    def test_backend(self):
        """The cProfile profiling backend should be used"""
        from yet_another_django_profiler.conf import settings
        assert settings.YADP_PROFILER_BACKEND == 'cProfile'


@pytest.mark.skipif(platform.python_implementation() != 'CPython' or
                    (sys.version_info[0] == 3 and sys.version_info[1] > 4),
                    reason='yappi does not yet work in this Python implementation')
@override_settings(YADP_ENABLED=True, YADP_PROFILER_BACKEND='yappi')
class YappiTest(TestCase, ParameterCases):
    """Profiling parameter tests using Yappi instead of cProfile"""

    def test_backend(self):
        """The Yappi profiling backend should be used"""
        from yet_another_django_profiler.conf import settings
        assert settings.YADP_PROFILER_BACKEND == 'yappi'


requirements/clean_up_requirements.py
#!/usr/bin/env python
# encoding: utf-8
# Created by Jeremy Bowman on Fri Jul 17 14:11:01 EDT 2015
# Copyright (c) 2015 Safari Books Online, LLC. All rights reserved.

"""
Utility to uninstall any installed packages listed in uninstall.txt.  Can't use
``pip uninstall -r`` because it aborts after the first package in the list
which isn't currently installed.  Also installs the currently-recommended
versions of setuptools and pip from requirements/nest.txt (to avoid needing
to replicate this logic in git-hooks/post-merge, tox.ini, quilter, etc.)
"""

from __future__ import print_function, unicode_literals

import codecs
import os
import re

from pip.download import PipSession
from pip.exceptions import UninstallationError
from pip.index import PackageFinder
from pip.req import parse_requirements


if __name__ == '__main__':
    requirements_dir = os.path.abspath(os.path.dirname(__file__))
    requirements_path = os.path.join(requirements_dir, 'uninstall.txt')

    session = PipSession()
    finder = PackageFinder([], [], session=session)
    requirements = parse_requirements(requirements_path, finder, session=session)
    print('Uninstalling former requirements...')
    for req in requirements:
        try:
            req.uninstall(auto_confirm=True)
            print('Uninstalled {}'.format(req.name))
        except UninstallationError as e:
            if 'not installed' in str(e):
                # Good, the whole point was that we don't want it installed
                pass
            else:
                # We probably want to hear about any other uninstallation problems
                raise
    print('Done')

    # Install the currently recommended setuptools and pip versions
    with codecs.open('requirements/base.txt', 'r', 'utf-8') as f:
        requirements = f.read()
    match = re.search(r'^setuptools==([\d\.]+)$', requirements, re.MULTILINE)
    os.system('pip install setuptools=={}'.format(match.group(1)))
    match = re.search(r'^pip==([\d\.]+)$', requirements, re.MULTILINE)
    os.system('pip install pip=={}'.format(match.group(1)))


yet_another_django_profiler/management/commands/profile.py
# encoding: utf-8
# Created by Jeremy Bowman on Fri Feb 21 11:16:36 EST 2014
# Copyright (c) 2014, 2015 Safari Books Online. All rights reserved.
#
# This software may be modified and distributed under the terms
# of the 3-clause BSD license.  See the LICENSE file for details.
"""
Yet Another Django Profiler "profile" management command
"""

from __future__ import unicode_literals

import atexit
import cProfile
import marshal
from optparse import make_option
import os
import pstats
import subprocess
import sys
import tempfile

try:
    from unittest import mock
except ImportError:
    import mock

import django
from django.conf import settings as django_settings
from django.core.management import call_command, ManagementUtility
from django.core.management.base import BaseCommand
from django.utils.six.moves import cStringIO as StringIO

from yet_another_django_profiler.conf import settings
from yet_another_django_profiler.middleware import func_strip_path, which

OPTIONS = (
    (('-o', '--output'), {'dest': 'path', 'help': 'Path to a file in which to store the profiling output (required if generating a call graph PDF, other results are output to the console by default)'}),
    (('-s', '--sort'), {'dest': 'sort', 'help': 'Statistic by which to sort the profiling data (default is to generate a call graph PDF instead)'}),
    (('-f', '--fraction'), {'dest': 'fraction', 'help': 'The fraction of total function calls to display (the default of .2 is omitted if max-calls or pattern are specified)'}),
    (('-m', '--max-calls'), {'dest': 'max_calls', 'help': 'The maximum number of function calls to display'}),
    (('-p', '--pattern'), {'dest': 'pattern', 'help': 'Regular expression filter for function display names'}),
    (('-b', '--backend'), {'dest': 'backend', 'default': 'cProfile', 'help': 'Profiler backend to use (cProfile or yappi)'}),
    (('-c', '--clock'), {'dest': 'clock', 'default': 'cpu', 'help': 'Yappi clock type to use (cpu or wall)'}),
)


class Command(BaseCommand):
    """
    Django management command for profiling other management commands.
    """
    args = 'other_command <argument argument ...>'
    help = 'Profile another Django management command'
    # Command line arguments for Django 1.7 and below
    custom_options = tuple([make_option(*option[0], **option[1]) for option in OPTIONS])

    @property
    def use_argparse(self):
        return not (django.VERSION[0] == 1 and django.VERSION[1] < 8)

    def add_arguments(self, parser):
        """Command line arguments for Django 1.8+"""
        for option in OPTIONS:
            parser.add_argument(*option[0], **option[1])

    def create_parser(self, prog_name, subcommand):
        """
        Override the base create_parser() method to ignore options of the
        command being profiled.
        """
        if not self.use_argparse:
            self.__class__.option_list = BaseCommand.option_list + self.custom_options
        parser = super(Command, self).create_parser(prog_name, subcommand)
        if not self.use_argparse:
            parser.disable_interspersed_args()
        return parser

    def handle(self, *args, **options):
        """
        Run and profile the specified management command with the provided
        arguments.
        """
        if not len(args):
            self.print_help(sys.argv[0], 'profile')
            sys.exit(1)
        if not options['sort'] and not options['path']:
            self.stdout.write('Output file path is required for call graph generation')
            sys.exit(1)

        command_name = args[0]
        utility = ManagementUtility(sys.argv)
        command = utility.fetch_command(command_name)
        parser = command.create_parser(sys.argv[0], command_name)
        if self.use_argparse:
            command_options = parser.parse_args(list(args[1:]))
            command_args = vars(command_options).pop('args', ())
        else:
            command_options, command_args = parser.parse_args(list(args[1:]))

        if command_name == 'test' and django_settings.TEST_RUNNER == 'django_nose.NoseTestSuiteRunner':
            # Ugly hack: make it so django-nose won't have nosetests choke on
            # our parameters
            BaseCommand.option_list += self.custom_options

        if options['backend'] == 'yappi' or (settings.YADP_PROFILER_BACKEND == 'yappi' and not options['backend']):
            import yet_another_django_profiler.yadp_yappi as yadp_yappi
            profiler = yadp_yappi.YappiProfile(wall=options['clock'] == 'wall')
        else:
            profiler = cProfile.Profile()

        if 'testing' not in options:
            atexit.register(output_results, profiler, options, self.stdout)
        profiler.runcall(call_command, command_name, *command_args,
                         stdout=self.stdout, **vars(command_options))
        if 'testing' in options:
            output_results(profiler, options, self.stdout)
        else:
            sys.exit(0)


def output_results(profiler, options, stdout):
    """Generate the profiler output in the desired format.  Implemented as a
    separate function so it can be run as an exit handler (because management
    commands often call exit() directly, bypassing the rest of the profile
    command's handle() method)."""
    profiler.create_stats()

    if not options['sort']:
        if not which('dot'):
            stdout.write('Could not find "dot" from Graphviz; please install Graphviz to enable call graph generation')
            return
        if not which('gprof2dot.py'):
            stdout.write('Could not find gprof2dot.py, which should have been installed by yet-another-django-profiler')
            return
        with tempfile.NamedTemporaryFile() as stats:
            stats.write(marshal.dumps(profiler.stats))
            stats.flush()
            cmd = ('gprof2dot.py -f pstats {} | dot -Tpdf'.format(stats.name))
            # Get a copy of the existing environment.
            env = os.environ.copy()
            # Add the PYTHONPATH using the existing python system path.
            env['PYTHONPATH'] = os.pathsep.join(sys.path)
            process = subprocess.Popen(cmd, shell=True, stdin=subprocess.PIPE,
                                       stdout=subprocess.PIPE, env=env)
            output = process.communicate()[0]
            return_code = process.poll()
            if return_code:
                stdout.write('gprof2dot/dot exited with {}'.format(return_code))
                return
            path = options['path']
            with open(path, 'wb') as pdf_file:
                pdf_file.write(output)
                stdout.write('Wrote call graph to {}'.format(path))
    else:
        sort = options['sort']
        if sort == 'file':
            # Work around bug on Python versions >= 2.7.4
            sort = 'fil'
        out = StringIO()
        stats = pstats.Stats(profiler, stream=out)
        with mock.patch('pstats.func_strip_path') as mock_func_strip_path:
            mock_func_strip_path.side_effect = func_strip_path
            stats.strip_dirs()
        restrictions = []
        if options['pattern']:
            restrictions.append(options['pattern'])
        if options['fraction']:
            restrictions.append(float(options['fraction']))
        elif options['max_calls']:
            restrictions.append(int(options['max_calls']))
        elif not options['pattern']:
            restrictions.append(.2)
        stats.sort_stats(sort).print_stats(*restrictions)
        if options['path']:
            path = options['path']
            with open(path, 'w') as text_file:
                text_file.write(out.getvalue())
                stdout.write('Wrote profiling statistics to {}'.format(path))
        else:
            stdout.write(out.getvalue())
